### PR TITLE
New metadata class TRestMapGroup

### PR DESCRIPTION
**Pull request by KonradAltenmueller**
KonradAltenmueller Large: 210 [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/konrad_map_group) [](https://github.com/rest-for-physics/framework/commits/konrad_map_group)

Changes proposed in this pull-request:

- new metadata class TRestMapGroup
- allows to define a std::map<string,string> which can be added to an analysis file to facilitate offline analysis or plotting. It can contain for example calibration factors for different signal IDs, or one can use it to assign names to signal IDs.

### CODE REVIEW

File: source/framework/core/inc/TRestMapGroup.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * If not, see http://www.gnu.org/licenses/.                             *
+ * For the list of contributors see $REST_PATH/CREDITS.                  *
+ *************************************************************************/
+
+#ifndef RestCore_TRestMapGroup
+#define RestCore_TRestMapGroup
+
+#include "TRestMetadata.h"
+
+/// The most simple metadata class to be used as a template
```
Comment (jgalan) [inline source/framework/core/inc/TRestMapGroup.h:28]: I guess the brief metadata class description needs an update

File: source/framework/core/inc/TRestMapGroup.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+class TRestMapGroup : public TRestMetadata {
+   protected:
+    /// Documentation of my dummy member
```
Comment (jgalan) [inline source/framework/core/inc/TRestMapGroup.h:31]: Members should be documented. Use `///` on the line above for each metadata member.

File: source/framework/core/src/TRestMapGroup.cxx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        groupDefinition = GetNextElement(groupDefinition);
+    }
+}
+
+///////////////////////////////////////////////
+/// Printing metadata member values on screen
+///
+void TRestMapGroup::PrintMetadata() {
+    TRestMetadata::PrintMetadata();
+
```
Comment (jgalan) [inline source/framework/core/src/TRestMapGroup.cxx:75]: What are the `fMapGroupKeys` and `fMapGroupValues` and why are not printed out?

File: source/framework/core/src/TRestMapGroup.cxx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+void TRestMapGroup::ConstructMap() {
+    for (unsigned int i = 0; i < fGroupKeys.size(); ++i) {
+        fMapGroup[fGroupKeys[i]] = fGroupValues[i];
+    }
+}
+///////////////////////////////////////////////
+/// \brief Set variables by default during initialization.
+///
+void TRestMapGroup::Initialize() { SetSectionName(this->ClassName()); }
+/// \brief Read Map Group from config file
```
Comment (jgalan) [inline source/framework/core/src/TRestMapGroup.cxx:58]: Format is inverted

**Review by jgalan — CHANGES_REQUESTED** [review]

**Comment by jgalan** [comment]
I thought the main purpose of this class would be to identify signal ids, i.e. integer numbers with a group name.

So, I would expect to have just a metadata member as:

```
std::map <string, std::vector<int>>;
```

where we store the group name, and the ids corresponding to that group.

The class then would be specific for signal ids, and perhaps the class could be renamed to something like `TRestMapGroupIds`.

**Comment by KonradAltenmueller** [comment]
I thought it is more general to have a `map<string, string>`. For example for calibration factors we have `map<int,double>`, for the veto names `map<string,int>` or `map<int,string>`... And it is no big deal to convert the string in whatever type we want later.

Anyways, I still have trouble to launch this class from an rml file and wanted to make a post in the rest forum about it. When I get it to work I will of course document all members and so on and address you other comments.

**Comment by jgalan** [comment]
Ok, but I think we are talking about two independent metadata classes. One to define sets of groups of ids, `TRestMapGroupIds` which I thought it was this PR --> `std::map< string, std::vector<int> >;` storing name + ids.

Then, a class to store the energy correction factors. This would be very similar to `TRestGainMap`. However, we do not implement a readout, so it would just serve to store correction factors.

```
std::map <int, double>;
```

File: source/framework/core/src/TRestMapGroup.cxx (hunk 3 above)
Comment (KonradAltenmueller) [reply source/framework/core/src/TRestMapGroup.cxx:126]: They are two vectors with the map keys and values, read from the RML. Now they are printed out, too.

File: source/framework/core/src/TRestMapGroup.cxx (hunk 4 above)
Comment (KonradAltenmueller) [reply source/framework/core/src/TRestMapGroup.cxx:109]: What do you mean?

**Comment by KonradAltenmueller** [comment]
I don't think it is necessary to separate the class in `TRestMapGroupIds` and `TRestMapFactors` (but maybe it is better to change the name of this class).
If I use this class as it is to obtain a `map<string, string> vetoGroup` by defining in the RML 
```
<TRestMapGroup name="vetoGroup">
      <MapGroup key="veto_top" value="1,2,3"/> 
</TRestMapGroup>
```
the vector of Signal IDs (1,2,3) for each veto group I can obtain with `REST_StringHelper::StringToElements(vetoGroup["veto_top"])`. 
Currently, this seems to be good enough to me... Or did you have something else in mind?

File: source/framework/core/src/TRestMapGroup.cxx (hunk 4 above)
Comment (jgalan) [reply source/framework/core/src/TRestMapGroup.cxx:109]: I mean, the `///////` should be before `/// \brief ....`

Comment (jgalan) [reply source/framework/core/src/TRestMapGroup.cxx:109]: The official format is as

```
////////////////////////
/// \brief ...
///
Int_t Method( xxx) {
```

Comment (jgalan) [reply source/framework/core/src/TRestMapGroup.cxx:109]: Leaving one blank line between methods

**Comment by jgalan** [comment]
I understand you may use `StringToElements`, but then why dont just use string variables, and create methods for every format? There is an adantage to use c++ types.

**Comment by jgalan** [comment]
> I don't think it is necessary to separate the class in `TRestMapGroupIds` and `TRestMapFactors` (but maybe it is better to change the name of this class). If I use this class as it is to obtain a `map<string, string> vetoGroup` by defining in the RML
> 
> ```
> <TRestMapGroup name="vetoGroup">
>       <MapGroup key="veto_top" value="1,2,3"/> 
> </TRestMapGroup>
> ```
> 
> the vector of Signal IDs (1,2,3) for each veto group I can obtain with `REST_StringHelper::StringToElements(vetoGroup["veto_top"])`. Currently, this seems to be good enough to me... Or did you have something else in mind?

Yes, but probably you need to implement a new method named `REST_StringHelper::StringToIntegerElements` so that you get back a `std::vector <int>`. You then fill the map with the group name and the vector. If the reply seems to be not accurate is because I didn't think about it with enough detail. Some times it happens to me.

**Pull request closed without merging**